app/controllers/home_controller.rb

```diff
@@ -73,12 +72,18 @@ def search_params
 
   def search_and_return
     if @search_form.valid?
-      begin
-        OnlineApplication.find_by!(reference: @search_form.reference.upcase)
-      rescue ActiveRecord::RecordNotFound
-        @search_form.errors.add(:reference, :not_found)
-        nil
+      @search = ApplicationSearch.new(@search_form.reference.upcase, current_user)
+      @matched = @search.for_hwf
+      if @matched.is_a? OnlineApplication
+        return @matched
+      else
+        @search_form.errors.add(:reference, @search.error_message)
       end
+      return nil
     end
   end
+
+  def can_show(application)
```

Comment:
This is not used anywhere, is it? If yes, same as bellow, we should use the `ApplicationPolicy`.